### ISSUE TITLE
Firestore: Incorrect implementation of collection.add()

### ISSUE THREAD

**Issue by samtstern** [googleapis/google-cloud-python#9629]
#### Environment details

Platforms: all

```
$ python --version
Python 2.7.17rc1
```

```
$ pip show google-cloud-firestore
Name: google-cloud-firestore
Version: 1.5.0
Summary: Google Cloud Firestore API client library
Home-page: https://github.com/GoogleCloudPlatform/google-cloud-python
Author: Google LLC
Author-email: googleapis-packages@google.com
License: Apache 2.0
Location: /usr/local/google/home/samstern/.local/lib/python2.7/site-packages
Requires: google-api-core, google-cloud-core, pytz
Required-by: firebase-admin
```

#### Steps to reproduce

This was originally reported to me here, and I reproduced it:
https://github.com/firebase/firebase-functions/issues/574

Python code:
```py
import firebase_admin
from firebase_admin import credentials
firebase_admin.initialize_app()
from google.cloud import firestore
db = firestore.Client()
print 'Adding'
db.collection('messages').add({'message':'Hello from Python'})
print 'Added'
```

Cloud Functions:
```js
const functions = require("firebase-functions");

exports.onNewMessageLength = functions.firestore
    .document("messages/{messageId}")
    .onCreate((snap, context) => {
        console.log("onCreate Data: ", JSON.stringify(snap.data()));
        return 0;
    });

exports.onUpdatedMessageLength = functions.firestore
    .document("messages/{messageId}")
    .onUpdate((change, context) => {
        console.log("onUpdate Data: ", JSON.stringify(change.after.data()));
        return 0;
    });
```

Running this in the Firebase emulators I see:
```
$ firebase emulators:exec "GCLOUD_PROJECT='fir-dumpster' python py/main.py && sleep 5"
⚠  Your requested "node" version "8" doesn't match your global version "10"
✔  functions: Emulator started at http://localhost:5001
i  firestore: Serving ALL traffic (including WebChannel) on http://localhost:8080
⚠  firestore: Support for WebChannel on a separate port (8081) is DEPRECATED and will go away soon. Please use port above instead.
i  firestore: Emulator logging to firestore-debug.log
✔  firestore: Emulator started at http://localhost:8080
i  firestore: For testing set FIRESTORE_EMULATOR_HOST=localhost:8080
i  functions: Watching "/tmp/tmp.OWPbChnnJ2/functions" for Cloud Functions...
✔  functions[onNewMessageLength]: firestore function initialized.
✔  functions[onUpdatedMessageLength]: firestore function initialized.
i  Running script: GCLOUD_PROJECT='fir-dumpster' python py/main.py && sleep 5
/usr/local/google/home/samstern/.local/lib/python2.7/site-packages/google/auth/_default.py:66: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK. We recommend that most server applications use service accounts instead. If your application continues to use end user credentials from Cloud SDK, you might receive a "quota exceeded" or "API not enabled" error. For more information about service accounts, see https://cloud.google.com/docs/authentication/.
  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)
Adding
Added
i  functions: Beginning execution of "onNewMessageLength"
>  onCreate Data:  {}
i  functions: Finished "onNewMessageLength" in ~1s
i  functions: Beginning execution of "onUpdatedMessageLength"
>  onUpdate Data:  {"message":{"type":"Buffer","data":[72,101,108,108,111,32,102,114,111,109,32,80,121,116,104,111,110]}}
i  functions: Finished "onUpdatedMessageLength" in ~1s
✔  Script exited successfully (code 0)
i  Shutting down emulators.
i  Stopping functions emulator
i  Stopping firestore emulator
```

What I should see is only the `onCreate` function firing and it should have the full data.  The implementation here is not correct:
https://github.com/googleapis/google-cloud-python/blob/master/firestore/google/cloud/firestore_v1/collection.py#L159-L180

**Comment by samtstern** [comment]
cc @BenWhitehead

**Comment by BenWhitehead** [comment]
@tseaver Can you take a look at this when you have a chance?

**Comment by tseaver** [comment]
Can we get this nailed down to something that doesn't involve `firesbase`, emulator output, or running anything on Cloud Functions?  E.g., if I run the Python script from above, what API calls do you expect it to make to the "real" backend API?  What state should be present?  Stripping down to stuff using only `google-cloud-firestore`:

```python
>>> import os
>>> from google.cloud.firestore_v1 import Client
>>> from google.oauth2 import service_account
>>> creds_file = os.environ['FIRESTORE_APPLICATION_CREDENTIALS']
>>> creds = service_account.Credentials.from_service_account_file(creds_file)
>>> client = Client(project=creds.project_id, credentials=creds)
>>> collection = client.collection('gcp-9629')
>>> doc_data = {'message': 'Reproduce https://github.com/googleapis/google-cloud-python/issues/9629'}
>>> timestamp, doc_ref = collection.add(doc_data)
>>> timestamp
seconds: 1573147518
nanos: 498621000

>>> doc_ref.path
'gcp-9629/V6XuLJ4fNIZizOcVYrxn'
>>> snapshot = doc_ref.get()
>>> snapshot.to_dict() == doc_data
True
>>> refs = list(collection.list_documents())
>>> refs[0] == doc_ref
True
```

That all looks correct to me, in terms of state visible from the outside:  we don't pass a document ID to `collection.add`, which means that it calls the `CreateDocument` API to get the server-generated document ID, and then calls the `Commit` API passing through the write PB to set the data.

Is the point of the issue to pass the document data along in the original `CreateDocument` API request, and avoid the `Commit` call?

**Comment by samtstern** [comment]
@tseaver `add()` is short for "locally generate a random document ID and then call `set()`"

The Node.js implementation is pretty clear:
https://github.com/googleapis/nodejs-firestore/blob/master/dev/src/reference.ts#L2129-L2133

So yes I believe what you want to do is pass the data into `_firestore_api.create_document()` and avoid any further API calls.

**Comment by samtstern** [comment]
Oh and in Node here's the implementation of `autoId()` for locally generating a document ID:
https://github.com/googleapis/nodejs-firestore/blob/master/dev/src/util.ts#L47-54

**Comment by tseaver** [comment]
@samtstern The call to `doc_ref.set()` is what triggers the `Commit` API call.  We can't pass the document data through to `_firestore_api.create_document`, because it might contain server transforms, which have to be processed in `Commit` (see issue #6862, fixed in PR #7072).

If generating a random document ID locally is OK / preferred, then we can avoid the call to the `CreateDocument` API, and just call `doc_ref.create`, which will invoke `Commit`.  We have a [comparable `_auto_id` function](https://github.com/googleapis/google-cloud-python/blob/79bdd03817879e495cae3534cbc168c3bf517ff2/firestore/google/cloud/firestore_v1/collection.py#L467-L474) already, which we [use in `Collection.document](https://github.com/googleapis/google-cloud-python/blob/79bdd03817879e495cae3534cbc168c3bf517ff2/firestore/google/cloud/firestore_v1/collection.py#L94-L111).

**Comment by samtstern** [comment]
@tseaver that makes sense.  Yes generating the random ID locally is preferred.
